### ISSUE TITLE
Annotations are not clipped properly

### ISSUE THREAD

**Issue by 2xB** [matplotlib/matplotlib#14354]
### Bug report

**Bug summary**

Drawing an annotation outside the visible region inline in jupyter notebook leads to whitespace added around the figure extending it to where the annotation would be.

**Code for reproduction**

```python
import matplotlib.pyplot as plt
%matplotlib inline
plt.xlim(0,1)
plt.ylim(0,1)
plt.annotate(xy=(3, 3), s='test')
```

**Actual outcome**

---
![grafik](https://user-images.githubusercontent.com/31772910/58486960-2098a500-8167-11e9-9ed2-3485d84e7c0d.png)

---

**Expected outcome**

---

![grafik](https://user-images.githubusercontent.com/31772910/58494641-dec32b00-8175-11e9-8fa8-98519c30c3e2.png)

---

**Matplotlib version**

  * Operating system: KDE neon 5.15
  * Matplotlib version: 3.0.3
  * Matplotlib backend (`print(matplotlib.get_backend())`): `module://ipykernel.pylab.backend_inline`
  * Python version: 3.6.7
  * Jupyter version: 5.2.2

Everything is installed via pip, pip and python are installed via apt.

**Comment by 2xB** [comment]
Notice that if one does not set xlim and ylim, the same happens, but with an other expected outcome, as discussed in #10497.

**Comment by ImportanceOfBeingErnest** [comment]
If the `clip_on` argument is set to `True`, 

    plt.annotate(xy=(3, 3), s='test', clip_on=True)

the output is as expected. I think it makes sense that the default `clip_on` is `False` for annotations, because often enough you do want to see the annotations, even if they are outside the axes. 

So I think the real bug here is that the annotation is invisible for the default `clip_on=False`.

**Comment by 2xB** [comment]
So this is purely an issue of setting defaults? I would then suggest to go from `clip_on=False, annotation_clip=True` to either `clip_on=True, annotation_clip=True` or `clip_on=False, annotation_clip=False`, because there is no reason to make place on the canvas for an annotation that is not shown.

**Comment by 2xB** [comment]
@ImportanceOfBeingErnest There is a catch: Why does this not occur when using `plt.savefig`?

```python
import matplotlib.pyplot as plt
plt.xlim(0,1)
plt.ylim(0,1)
plt.annotate(xy=(3, 3), s='test')
plt.savefig("test.png")
```

---

![test2](https://user-images.githubusercontent.com/31772910/58546206-2b097c00-8205-11e9-9d3b-c35da2f9d516.png)

---

**Comment by ImportanceOfBeingErnest** [comment]
No, it's not an issue of the correct defaults. The issue is that `annotation_clip` is considered only for finding out whether or not to *draw* the annotation, but not for finding out if it is to be taken into account for the calculation of the tight_bbox. 

The issue only occurs if you use `plt.savefig("test.png", bbox_inches="tight")` (which is what the jupyter/IPython inline backend does by default). 

In principle the defaults are fine: `clip_on=False` and `annotation_clip=None`. Those ensure that annotations that are defined in coordinates other than data coordinates are still shown outside the axes, which is what most people would expect. 

So possible solutions are:

* In the tight_bbox calculations, check if an artist is an annotation, or has a `_check_xy` attribute, and include it or not depending on its return. Drawback of this: It makes drawing less efficient and is in principle bad style, because in all other cases we would rather ask the artist if it to be included or not through the `clip_on` property.
* Synchronize the `clip_on` with the `annotation_clip`. Meaning, `clip_on` might change its state depending on the current value of `annotation_clip` and  the coordinate system in use. This sounds a lot better, but is harder to implement. Also it's not clear which one should take precedence in case both are defined.

**Comment by 2xB** [comment]
As of my understanding while quickly looking through the source, lib/matplotlib/artist.py uses `clip_on` in https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/artist.py#L268, so more or less any value for clip_on other than None or False is True. Therefore, wouldn't it be possible to introduce a new state to `clip_on`, so it stores `clip_on=True`, `clip_on=False` or `clip_on=Annotation.ANNOTATION_CLIP`, where the latter replaces `annotation_clip=True`? Then you could not have a weird combination of `clip_on` and `annotation_clip`, as their functionality is combined into one parameter.
This idea probably needs polishment, but it would help not having to synchronize multiple parameters.

**Comment by ImportanceOfBeingErnest** [comment]
Due to the drawbacks of both possible solutions to this, I don't think this will be solved anytime soon. Also, one may consider it as a documentation issue, clarifying the role of `clip_on` and `annotation_clip`. 

Hence, relabeling and remilestoning.

**Comment by 2xB** [comment]
> In principle the defaults are fine: `clip_on=False` and `annotation_clip=None`. Those ensure that annotations that are defined in coordinates other than data coordinates are still shown outside the axes, which is what most people would expect.

Sorry for reacting so late to this, but in my issue, the annotation is NOT shown outside the axes, although I do not change the defaults.

**Comment by ImportanceOfBeingErnest** [comment]
Yes, but you can easily get the annotation to show up using 

    plt.annotate(xy=(3, 3), s='test', annotation_clip=False)

So the question is, whether it's worth putting in a hacky solution with some heuristics about whether or not to show the annotation in case `bbox_inches="tight"` is used.

**Comment by jklymak** [comment]
```
annotation_clip : bool or None, optional
    Whether to draw the annotation when the annotation point *xy* is
    outside the axes area.

    - If *True*, the annotation will only be drawn when *xy* is
      within the axes.
    - If *False*, the annotation will always be drawn.
    - If *None*, the annotation will only be drawn when *xy* is
      within the axes and *xycoords* is 'data'.
```

So, by default its none, and its not drawn.  But because its not clipped using the usual clipping argument, its still included in the tight_layout calculation.  I'm not sure I understand the subtleties of why annotations have multiple clip arguments, or what behaviour is particularly desired here.

**Comment by 2xB** [comment]
@ImportanceOfBeingErnest I'm not quite sure I get this. When the default for `annotation_clip` is not to show the annotation outside plot borders, should not then the default for `clip_on` be to clip outside plot borders? Not to show the annotation outside but reserving space for it seems like a confusing default.
This would also improve consistency with every other plot item (like scatterplots, line plots etc.), that are clipped outside plot borders.

**Comment by ImportanceOfBeingErnest** [comment]
The point is, you *do* want to show the annotations outside the axes in cases the anchor point is in data coordinates and still inside the axes, or in cases where you have used coordinates, other than data coordinates. Maybe the below animation makes this clearer.

![annotation](https://user-images.githubusercontent.com/23121882/63422669-1e16b600-c40b-11e9-8c3d-6a2e956ae670.gif)

**Comment by 2xB** [comment]
Thank you for the explanation and the work of doing a GIF!
I think I might understand the reasoning, and have a suggestion on how to fix this. To discuss it, I created PR #15096.

**Comment by ImportanceOfBeingErnest** [comment]
In #15096 @jklymak asked 
> Is there any use to annotation_clip and clip_on to ever be different?

As answer, here are all combinations:

![image](https://user-images.githubusercontent.com/23121882/63465165-efc3c580-c460-11e9-87a8-bf400020030d.png)

**Comment by jklymak** [comment]
OK, but I don't think we *ever* want the `clip_on=True` result.  Is there a practical use for that?

**Comment by ImportanceOfBeingErnest** [comment]
I'm not sure. Maybe in cases where you don't have any arrow? What's the alternative, though? Giving `clip_on` a third state `None` that is only used by annotations? But that would still require to somehow special-casing annotations in the bbox_inches="tight"- algorithms, right?

**Comment by fperonaci** [comment]
Hi, just to give a feedback (I hope this is the appropriate place) I agree with @jklymak: the case `clip_on=True` is weird and, I would argue, actually harmful. Indeed, I thought that the behaviour of `annotation_clip=True` would be achieved with `clip_on=True` which I think is the reasonable expectation. I think it makes a lot of sense to consider the annotation as a whole.

**Comment by rcomer** [comment]
I do not know what changed but I just tried this with Matplotlib v3.8.4 in a notebook and got the expected result, i.e. the figure is **not** resized to include the invisible annotation.
```python
import matplotlib.pyplot as plt

plt.xlim(0,1)
plt.ylim(0,1)
plt.annotate('test', xy=(3, 3))
```

**Comment by QuLogic** [comment]
This was fixed by #18772.
